### PR TITLE
Api register

### PR DESCRIPTION
**Pull request by Jarod42**
**What does this PR do?**

Move api.register in a more logical place

**How does this PR change Premake's behavior?**

No changes

**Did you check all the boxes?**

- [x] Focus on a single fix or feature; remove any unrelated formatting or code changes
- [ ] Add unit tests showing fix or feature works; all tests pass
- [ ] Mention any [related issues](https://github.com/premake/premake-core/issues) (put `closes #XXXX` in comment to auto-close issue when PR is merged)
- [x] Follow our [coding conventions](https://github.com/premake/premake-core/blob/master/CONTRIBUTING.md#coding-conventions)
- [ ] Minimize the number of commits
- [ ] Align [documentation](https://github.com/premake/premake-core/tree/master/website) to your changes

### CODE REVIEW

**Comment by nickclark2016** [comment]
> Nice cleanup.
> 
> Some other ideas for some further cleanups around this area.
> 
> 1. Maybe moving all remaining API-related calls into a new `_premake_api.lua`
> 2. Maybe introduce a prefix for all toolset and action-specific APIs, so `gcc.prefix`, `android.fpu`, `vs.*` and soft-deprecate all existing ones. This is a fairly big change, but think it would be a good future improvement.
> 3. Move all toolset or action-specific APIs to respective named folders in the website docs.

1. This is a good idea. I would be in favor of this.
2. This is definitely a bigger change, so I'd want to get input from more of the team. I'm not sure where I fall on this one.
3. I'm against this idea, because it would require a refactor every time we add support for an API in an additional action or toolset.

File: src/tools/dotnet.lua [hunk 1]
```diff
@@ -11,6 +11,21 @@
 	local project = p.project
 	local config = p.config
 
+	p.api.register {
+		name = "csversion",
+		scope = "config",
+		kind = "string",
+	}
+
+	p.api.register {
```
[the remark this answers is not in the document]
Comment (Jarod42) [reply src/tools/dotnet.lua:20]: I agree, adding doc in outside scope of this PR

**Comment by Jarod42** [comment]
> > Nice cleanup.
> > Some other ideas for some further cleanups around this area.
> > 
> > 1. Maybe moving all remaining API-related calls into a new `_premake_api.lua`
>
> 1. This is a good idea. I would be in favor of this.

I would prefer to do it in another PR.
(minimize conflicts opportunity, and simplify resolution)

> > 2. Maybe introduce a prefix for all toolset and action-specific APIs, so `gcc.prefix`, `android.fpu`, `vs.*` and soft-deprecate all existing ones. This is a fairly big change, but think it would be a good future improvement.
>
> 2. This is definitely a bigger change, so I'd want to get input from more of the team. I'm not sure where I fall on this one.

This one in a (big) API change, so definitively in another PR.
and require approval of the idea before.
BTW, it would solve issue with conflicting API from different module I already encounter.

> > 3. Move all toolset or action-specific APIs to respective named folders in the website docs.
>
> 3. I'm against this idea, because it would require a refactor every time we add support for an API in an additional action or toolset.

I see that as followup on this PR, but outside the scope of this PR.
I (try to) move API not from support, but from "specificity",
so an API which is only supported by visual, but which makes sense for other generator goes in generic API, not visual one.

`gcc.prefix` is purely gcc specific.
`startproject` (only supported by visual) should stay in generic as it make sense for (some) other generator.

Some API placement might be discutable, I kept `nuget` which targets visual/.Net in generic whereas it might make sense to move it in visual API.
gcc/clang (i.e no msc) specific goes in generic whereas a group might make sense

**Comment by Jarod42** [comment]
> @Jarod42 This is now reporting as conflicted btw.

@tritao : Thanks, rebased (was on 'Remove "empty" api.' commit (No conflicts with perforce-merge :-) )

**Comment by nickclark2016** [comment]
Is this something that absolutely must make it into beta3? @Jarod42

**Review by nickclark2016 — APPROVED** [review]

**Pull request merged**